File: DotNet/PopulationFitness/PopulationFitness/Models/GenerationStatistics.cs
using System;
using System.Collections.Generic;

namespace PopulationFitness.Models
{
    /**
     * Created by pete.callaghan on 04/07/2017.
     */
    public class GenerationStatistics
    {
        public readonly int NumberBorn;
        public readonly int NumberKilled;
        public readonly int Population;
        public readonly int Year;
        public readonly Epoch Epoch;
        public readonly long BornTime;
        public readonly long KillTime;
        public double AverageFitness;
        public double AverageFactoredFitness;
        public double FitnessDeviation;
        public double AverageAge;
        public double CapacityFactor;
        public double AverageMutations;
        public double AverageLifeExpectancy;

        public GenerationStatistics(Epoch epoch,
                                    int year,
                                    int population,
                                    int number_born,
                                    int number_killed,
                                    long born_time,
                                    long kill_time,
                                    double capacity_factor,
                                    double average_mutations)
        {
            NumberBorn = number_born;
            NumberKilled = number_killed;
            Population = population;
            Year = year;
            Epoch = epoch;
            BornTime = born_time;
            KillTime = kill_time;
            AverageFitness = 0;
            AverageFactoredFitness = 0;
            FitnessDeviation = 0;
            AverageAge = 0;
            CapacityFactor = capacity_factor;
            AverageMutations = average_mutations;
            AverageLifeExpectancy = 0;
        }

        public double BornElapsedInHundredths()
        {
            return (double)(BornTime / 10) / 100;
        }

        public double KillElapsedInHundredths()
        {
            return (double)(KillTime / 10) / 100;
        }

        private static long Average(long value, long count)
        {
            if (count < 1)
            {
                return value;
            }
            return value / count;
        }

        private static double Average(double value, long count)
        {
            if (count < 1)
            {
                return value;
            }
            return value / count;
        }

        public static GenerationStatistics Add(GenerationStatistics first, GenerationStatistics second)
        {
            if (first.Year != second.Year) throw new Exception("Cannot add different years");

            GenerationStatistics result = new GenerationStatistics(new Epoch(first.Epoch),
                    first.Year,
                    first.Population + second.Population,
                    first.NumberBorn + second.NumberBorn,
                    first.NumberKilled + second.NumberKilled,
                    first.BornTime + second.BornTime,
                    first.KillTime + second.KillTime,
                    first.CapacityFactor,
                    first.AverageMutations);
            result.AverageAge = Average(first.AverageAge * first.Population + second.AverageAge * second.Population, result.Population);
            result.AverageFitness = Average(first.AverageFitness * first.Population + second.AverageFitness * second.Population, result.Population);
            result.AverageFactoredFitness = Average(first.AverageFactoredFitness * first.Population + second.AverageFactoredFitness * second.Population, result.Population);
            result.Epoch.ExpectedMaxPopulation += second.Epoch.ExpectedMaxPopulation;
            result.Epoch.EnvironmentCapacity += second.Epoch.EnvironmentCapacity;
            result.CapacityFactor = Average(first.CapacityFactor * first.Population + second.CapacityFactor * second.Population, result.Population);
            result.AverageMutations = Average((first.AverageMutations * first.NumberBorn + second.AverageMutations * second.NumberBorn), result.NumberBorn);
            result.AverageLifeExpectancy = Average(first.AverageLifeExpectancy * first.NumberKilled + second.AverageLifeExpectancy * second.NumberKilled, result.NumberKilled);
            return result;
        }

        public static List<GenerationStatistics> Add(List<GenerationStatistics> first, List<GenerationStatistics> second)
        {
            List<GenerationStatistics> result = new List<GenerationStatistics>();

            if (first.Count != second.Count)
            {
                throw new Exception("Cannot add different numbers of generations");
            }

            for (int i = 0; i < first.Count; i++)
            {
                result.Add(Add(first[i], second[i]));
            }

            return result;
        }
    }
}

File: Scala/PopulationFitness/src/main/scala/uk/edu/populationfitness/models/history/GenerationStatistics.scala
package uk.edu.populationfitness.models.history

import uk.edu.populationfitness.models.Population.{Born, Change, Killed}
import uk.edu.populationfitness.models.history.GenerationStatistics.average

object GenerationStatistics {
  private def average(value: Double, count: Long): Double = {
    if (count < 1) value else value / count
  }

  def add(first: Seq[GenerationStatistics], second: Seq[GenerationStatistics]): Seq[GenerationStatistics] = {
    (first, second).zipped.map(_ + _)
  }

  def apply(epoch: Epoch, year: Int, latest: Change with Born with Killed): GenerationStatistics = {
    val pop = latest.population
    new GenerationStatistics(epoch,
      year,
      latest.population.size,
      latest.born.size,
      latest.killed.size,
      latest.bornElapsed,
      latest.killedElapsed,
      pop.capacityFactor,
      pop.averageMutations)
    {
      averageFitness = pop.averageFitness
      averageFactoredFitness = pop.averageFactoredFitness
      fitnessDeviation = pop.standardDeviationFitness
      averageAge = pop.averageAge(year)
      averageLifeExpectancy = latest.killed.averageAge(year)
    }
  }
}

class GenerationStatistics(val epoch: Epoch,
                           val year: Int,
                           val population: Int,
                           val numberBorn: Int,
                           val numberKilled: Int,
                           val bornTime: Long,
                           val killTime: Long,
                           var capacityFactor: Double,
                           var averageMutations: Double) {
  var averageFitness = .0
  var averageFactoredFitness = .0
  var fitnessDeviation = .0
  var averageAge = .0
  var averageLifeExpectancy = .0

  def bornElapsedInHundredths: Double = (bornTime / 10).toDouble / 100

  def killElapsedInHundredths: Double = (killTime / 10).toDouble / 100

  def +(g: GenerationStatistics): GenerationStatistics = {
    if (year != g.year) throw new Error("Cannot add different years")
    val result = new GenerationStatistics(new Epoch(epoch),
      year,
      population + g.population,
      numberBorn + g.numberBorn,
      numberKilled + g.numberKilled,
      bornTime + g.bornTime,
      killTime + g.killTime,
      capacityFactor,
      averageMutations)

    result.averageAge = average(averageAge * population + g.averageAge * g.population, result.population)
    result.averageFitness = average(averageFitness * population + g.averageFitness * g.population, result.population)
    result.averageFactoredFitness = average(averageFactoredFitness * population + g.averageFactoredFitness * g.population, result.population)
    result.epoch.expectedMaxPopulation += g.epoch.expectedMaxPopulation
    result.epoch.environmentCapacity += g.epoch.environmentCapacity
    result.capacityFactor = average(capacityFactor * population + g.capacityFactor * g.population, result.population)
    result.averageMutations = average(averageMutations * numberBorn + g.averageMutations * g.numberBorn, result.numberBorn)
    result.averageLifeExpectancy = average(averageLifeExpectancy * numberKilled + g.averageLifeExpectancy * g.numberKilled, result.numberKilled)
    result
  }
}
